# Australia Vehicle Price Prediction

By - Abhinav Sikhwal

This is a Dataset obtained from Kaggle for the purpose of understanding and designing Machine Learning predictions for the M505B Course at Gisma University of Applied Sciences.

Dataset URL: https://www.kaggle.com/datasets/nelgiriyewithana/australian-vehicle-prices 

# Problem Statement

The Data contains valuable Information about Different attributes of Used / New Cars from Australia with their Selling Price. As the Dataset contains very spefic information about the age, make, model of Cars it would be ideal to Explore, Process and Engineer the Data so as to be able to try Machine Learning Algorithms.

Since this seems like a Supervised Learning Regression Problem as we have the target variable of the price this Notebook will attempt to train different Models on the Dataset to find the Alrorithm which best captures the relationship between the Variables.

# Dataset Information

This Dataset contains a range of Values and the following Columns :

Brand, 
Year, 
Model, 
Car / Suv, 
Title, 
Used Or New, 
Type of transmission (manual or automatic), 
Engine, 
Drive Type,
Fuel Type, 
Fuel Consumption, 
Kilometres, 
Colour(Exterior and Interior), 
Location, 
Number of Cylinders in Engine, 
Body Type, 
Number of Doors, 
Number of Seats,
Price of the car in Australian dollars. 

# Importing the Useful Libraries

In [86]:
import pandas as pd
import numpy as np
import re
from sklearn.model_selection import train_test_split
import sklearn.preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Loading the Dataset

In [87]:
# Loading the dataset using pandas
car_df = pd.read_csv("Australian Vehicle Prices.csv")

pd.set_option('display.max_columns', None)

In [88]:
# To see the first few columns
car_df.head(10)

Brand    Year   Model               Car/Suv  \
0      Ssangyong  2022.0  Rexton  Sutherland Isuzu Ute   
1             MG  2022.0     MG3             Hatchback   
2            BMW  2022.0    430I                 Coupe   
3  Mercedes-Benz  2011.0    E500                 Coupe   
4        Renault  2022.0  Arkana                   SUV   
5         Toyota  2004.0  Estima        ON FOUR WHEELS   
6           Land  2017.0   Rover                   SUV   
7         Nissan  2000.0  Pulsar             Hatchback   
8         Toyota  2013.0      86                 Coupe   
9          Honda  2014.0    Jazz             Hatchback   

                                               Title UsedOrNew Transmission  \
0               2022 Ssangyong Rexton Ultimate (awd)      DEMO    Automatic   
1          2022 MG MG3 Auto Excite (with Navigation)      USED    Automatic   
2                              2022 BMW 430I M Sport      USED    Automatic   
3                   2011 Mercedes-Benz E500 Elegance      USED    Automatic   
4                         2022 Renault Arkana Intens      USED    Automatic   
5                       2004 Toyota Estima T EDITION      USED    Automatic   
6  2017 Land Rover Range Rover Evoque TD4 180 HSE...      USED    Automatic   
7                              2000 Nissan Pulsar LX      USED    Automatic   
8                                  2013 Toyota 86 GT      USED    Automatic   
9                             2014 Honda Jazz Hybrid      USED    Automatic   

         Engine DriveType  FuelType FuelConsumption Kilometres   ColourExtInt  \
0  4 cyl, 2.2 L       AWD    Diesel  8.7 L / 100 km       5595  White / Black   
1  4 cyl, 1.5 L     Front   Premium  6.7 L / 100 km         16  Black / Black   
2    4 cyl, 2 L      Rear   Premium  6.6 L / 100 km       8472   Grey / White   
3  8 cyl, 5.5 L      Rear   Premium   11 L / 100 km     136517  White / Brown   
4  4 cyl, 1.3 L     Front  Unleaded    6 L / 100 km       1035   Grey / Black   
5             -     Other  Unleaded               -     160230       Grey / -   
6    4 cyl, 2 L       AWD    Diesel  5.1 L / 100 km      67662  White / Black   
7  4 cyl, 1.6 L     Front  Unleaded    8 L / 100 km     300539     Red / Grey   
8    4 cyl, 2 L      Rear   Premium  7.1 L / 100 km      82012      Black / -   
9  4 cyl, 1.3 L     Front    Hybrid  4.5 L / 100 km      38229       Blue / -   

              Location CylindersinEngine    BodyType     Doors     Seats  \
0       Caringbah, NSW             4 cyl         SUV   4 Doors   7 Seats   
1       Brookvale, NSW             4 cyl   Hatchback   5 Doors   5 Seats   
2        Sylvania, NSW             4 cyl       Coupe   2 Doors   4 Seats   
3    Mount Druitt, NSW             8 cyl       Coupe   2 Doors   4 Seats   
4     Castle Hill, NSW             4 cyl         SUV   4 Doors   5 Seats   
5     Saint Marys, NSW                 -  Commercial       NaN       NaN   
6       Blacktown, NSW             4 cyl         SUV   2 Doors   4 Seats   
7  Wentworthville, NSW             4 cyl   Hatchback   5 Doors   5 Seats   
8   Mcgraths Hill, NSW             4 cyl       Coupe   2 Doors   4 Seats   
9        Lidcombe, NSW             4 cyl   Hatchback   5 Doors   5 Seats   

    Price  
0   51990  
1   19990  
2  108988  
3   32990  
4   34990  
5    9990  
6   62280  
7    2995  
8   24888  
9   17900

In [89]:
# To check the size of the dataset
print(" The Size of the Dataset is :", car_df.shape)

 The Size of the Dataset is : (16734, 19)


In [90]:
#Checking the Datatypes of all the Columns
car_df.dtypes

Brand                 object
Year                 float64
Model                 object
Car/Suv               object
Title                 object
UsedOrNew             object
Transmission          object
Engine                object
DriveType             object
FuelType              object
FuelConsumption       object
Kilometres            object
ColourExtInt          object
Location              object
CylindersinEngine     object
BodyType              object
Doors                 object
Seats                 object
Price                 object
dtype: object

In [91]:
# Checking more information about the dataset
car_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16734 entries, 0 to 16733
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Brand              16733 non-null  object 
 1   Year               16733 non-null  float64
 2   Model              16733 non-null  object 
 3   Car/Suv            16706 non-null  object 
 4   Title              16733 non-null  object 
 5   UsedOrNew          16733 non-null  object 
 6   Transmission       16733 non-null  object 
 7   Engine             16733 non-null  object 
 8   DriveType          16733 non-null  object 
 9   FuelType           16733 non-null  object 
 10  FuelConsumption    16733 non-null  object 
 11  Kilometres         16733 non-null  object 
 12  ColourExtInt       16733 non-null  object 
 13  Location           16284 non-null  object 
 14  CylindersinEngine  16733 non-null  object 
 15  BodyType           16452 non-null  object 
 16  Doors              151

In [92]:
#Checking for Null values in the dataset
car_df.isna().sum()

Brand                   1
Year                    1
Model                   1
Car/Suv                28
Title                   1
UsedOrNew               1
Transmission            1
Engine                  1
DriveType               1
FuelType                1
FuelConsumption         1
Kilometres              1
ColourExtInt            1
Location              450
CylindersinEngine       1
BodyType              282
Doors                1604
Seats                1705
Price                   3
dtype: int64

# Data Exploration

The Dataset has more than 16000 rows and 19 Columns. It contains some Null Values that need to be processed. Aside from the Year Column, all other columns have a Data type of Object which is not ideal for Machine Learning.

In [93]:
#Checking unique values in the columns
print(pd.unique(car_df['UsedOrNew']))

['DEMO' 'USED' 'NEW' nan]


In [94]:
print(pd.unique(car_df['Transmission']))

['Automatic' 'Manual' '-' nan]


In [95]:
print(pd.unique(car_df['DriveType']))

['AWD' 'Front' 'Rear' 'Other' '4WD' nan]


In [96]:
print(pd.unique(car_df['BodyType']))

['SUV' 'Hatchback' 'Coupe' 'Commercial' nan 'Wagon' 'Ute / Tray' 'Sedan'
 'People Mover' 'Convertible' 'Other']


In [97]:
print(pd.unique(car_df['FuelType']))

['Diesel' 'Premium' 'Unleaded' 'Hybrid' '-' 'Other' 'Electric' 'LPG'
 'Leaded' nan]


In [98]:
print(pd.unique(car_df['Brand']))

['Ssangyong' 'MG' 'BMW' 'Mercedes-Benz' 'Renault' 'Toyota' 'Land' 'Nissan'
 'Honda' 'Volkswagen' 'Mitsubishi' 'Ford' 'Subaru' 'Hyundai' 'Jeep'
 'Volvo' 'Mazda' 'Abarth' 'Holden' 'Audi' 'Kia' 'Mini' 'Peugeot' 'Suzuki'
 'Dodge' 'Porsche' 'Daihatsu' 'Isuzu' 'Lexus' 'Jaguar' 'Rolls-Royce'
 'Skoda' 'Fiat' 'Haval' 'Citroen' 'LDV' 'HSV' 'Foton' 'Mahindra'
 'Maserati' 'Alfa' 'GWM' 'Ram' 'Tesla' 'Genesis' 'Chrysler' 'Great' 'Opel'
 'Saab' 'Bentley' 'Rover' nan 'Ferrari' 'Cupra' 'Chevrolet' 'Lamborghini'
 'FPV' 'McLaren' 'Chery' 'Iveco' 'Hummer' 'Infiniti' 'BYD' 'Tata' 'Aston'
 'Daewoo' 'Proton' 'Geely' 'Smart' 'Ineos' 'Caterham' 'Hino'
 'International' 'Lotus' 'Polestar' 'Datsun' 'Packard']


In [99]:
print(pd.unique(car_df['Doors']))

[' 4 Doors' ' 5 Doors' ' 2 Doors' nan ' 3 Doors' ' 7 Seats' ' 5 Seats'
 ' 4 Seats' ' 8 Seats' ' 2 Seats' ' 9 Seats' ' 6 Seats' ' 3 Seats'
 ' 12 Seats']


In [100]:
print(pd.unique(car_df['Seats']))

[' 7 Seats' ' 5 Seats' ' 4 Seats' nan ' 2 Seats' ' 3 Seats' ' 14 Seats'
 ' 8 Seats' ' 6 Seats' ' 9 Seats' ' 11 Seats' ' 12 Seats' ' 22 Seats'
 ' 15 Seats']


In [101]:
print(pd.unique(car_df['CylindersinEngine']))

['4 cyl' '8 cyl' '-' '3 cyl' '6 cyl' '5 cyl' '12 cyl' '0 L' '2 cyl' nan
 '10 cyl' '2 L']


# Data Insights

Now that we are done with Exploring the Dataset there are a few insights:

The Brand Names have a lot of Categorical Values of Maker names,

The Engine Column is format of "4.0 / 2 L",

The Color Column is in Format of "White / Black",

The Cylinders is in Format of " 6 cyl ",

The Doors and Seats columns are in Format of " 6 Doors",

These Columns need to be separated and converted to their corresponding Numerical attributes.

# Data Preprocessing

Most of the rows are in object form and need to be converted to Integer / Float for better processing by the machine learning algorithms.

For this purpose, using functions in python we will extract the categorical values and convert them to numerical values using formulas. 

In [102]:
# To Remove the "-", "POA" as well as other Null Values in the DataSet
car_df.replace('-', pd.NA, inplace=True)
car_df.replace('POA', pd.NA, inplace=True)
car_df = car_df.dropna()

In [103]:
# Splitting the string and extracting the numeric part and converting to Float to Divide by 100

def newfuel(fuel):
    cons_str = fuel.split()[0]
    cons_float = float(cons_str)
    return cons_float / 100

# Applying the function to the fuel consumption column of the Dataset
car_df['Fuel efficiency (L/km)'] = car_df['FuelConsumption'].apply(newfuel)

In [104]:
# Converting Kilometres and Price to Float Values
car_df['Kilometres'] = car_df['Kilometres'].astype(float)
car_df['Price'] = car_df['Price'].astype(float)

# Converting Doors and Seats to Float Values
car_df['Doors'] = car_df['Doors'].str.extract(r'(\d+)').astype(float)
car_df['Seats'] = car_df['Seats'].str.extract(r'(\d+)').astype(float)

In [105]:
#Converting the Cylinders Column to extract the Engine Capacity into a New Column

car_df['CylindersinEngine'] = car_df['CylindersinEngine'].str.extract('(\d+)').astype(float)

def cylinder(text):
    match = re.search(r'(\d+\.\d+|\d+) L', text)
    if match:
        return float(match.group(1))
    else:
        return None

car_df['Engine'] = car_df['Engine'].apply(cylinder)

In [106]:
# Separating the Interior and Exterior Color Values
car_df[['Interior', 'Exterior']] = car_df['ColourExtInt'].str.extract(r'^(.*) / (.*)$')

In [107]:
# To consider the DEMO car as a NEW car
car_df['UsedOrNew'] = car_df['UsedOrNew'].replace('DEMO', 'NEW')

# Feature Engineering

Since most of the features / columns of this dataset are in Categorical form, they need to processed and converted to numerical data for better understanding by the machine learning algorithms. 

The Values will then be encoded using Label Encoding which will convert the category to a numeircal number.

or One Hot Encoding which will create a binary representation of the diferent categories. 

Additionaly target encoding will also be used for some columns. 

# Scaling the Age and Kilometres Value

The Year Column has been used to create the Age Column.

The Age and Kilometres Column will now be Scaled between the values of 0 and 1 using Min max scaler. However since the more the age and kilometres , the worse the car they will be scaled in reverse order.

In [108]:
car_df['age'] = 2024 - car_df['Year']

scaler = MinMaxScaler()

car_df['age'] = 1 - scaler.fit_transform(car_df[['age']])
car_df['Kilometres'] = 1 - scaler.fit_transform(car_df[['Kilometres']])

# Columns for Target Encoding

Model,
Car / Suv,
Drive type,
Body Type,

In [109]:
target_encode = ['Model', 'Car/Suv', 'DriveType','BodyType','Interior','Exterior']

for col in target_encode:
    if col != 'Price':
        means = car_df.groupby(col)['Price'].mean()
        car_df[col + '_encoded'] = car_df[col].map(means)

# Encoding - One Hot Encoding and Label Encoding

Some of the Columns that have very few Unique Values (UsedorNew, Transmission, Fueltype) will be encoded using One hot encoding. 

The rest of the Categorical Columns will be encoded using Label Encoding. This is to Avoid Dimensionality of the Data.

In [110]:
# One Hot Encoding
cardf = pd.get_dummies(car_df, columns=['UsedOrNew', 'Transmission', 'FuelType'])

In [111]:
# Label Encoding

label_encoder = LabelEncoder()
cardf['brand'] = label_encoder.fit_transform(car_df['Brand'])
cardf['Enginec'] = label_encoder.fit_transform(car_df['Engine'])
cardf['Cylinder'] = label_encoder.fit_transform(car_df['CylindersinEngine'])

In [112]:
# Now to Remove the Unnecessary Columns and only keeping the transformed columns

cardf.drop(columns = ['Brand','Engine','CylindersinEngine','Model','Title','Year','Car/Suv','DriveType','FuelConsumption','Interior','Exterior','ColourExtInt','Location','BodyType'], inplace = True)

cardf.head()

Kilometres  Doors  Seats     Price  Fuel efficiency (L/km)       age  \
0    0.989368    4.0    7.0   51990.0                   0.087  0.977778   
1    0.999971    5.0    5.0   19990.0                   0.067  0.977778   
2    0.983900    2.0    4.0  108988.0                   0.066  0.977778   
3    0.740543    2.0    4.0   32990.0                   0.110  0.733333   
4    0.998035    4.0    5.0   34990.0                   0.060  0.977778   

   Model_encoded  Car/Suv_encoded  DriveType_encoded  BodyType_encoded  \
0   48712.916667     51990.000000       46817.365238      37885.964618   
1   18651.541667     22304.407565       25749.832049      22646.521937   
2   55504.600000     67613.940945       36512.875189      67642.466667   
3   25990.000000     67613.940945       36512.875189      67642.466667   
4   42046.866667     36286.875869       25749.832049      37885.964618   

   Interior_encoded  Exterior_encoded  UsedOrNew_NEW  UsedOrNew_USED  \
0      35474.428899      37295.919777           True           False   
1      39698.609202      37295.919777          False            True   
2      38393.810176      59319.411765          False            True   
3      35474.428899      71020.291667          False            True   
4      38393.810176      37295.919777          False            True   

   Transmission_Automatic  Transmission_Manual  FuelType_Diesel  \
0                    True                False             True   
1                    True                False            False   
2                    True                False            False   
3                    True                False            False   
4                    True                False            False   

   FuelType_Electric  FuelType_Hybrid  FuelType_LPG  FuelType_Leaded  \
0              False            False         False            False   
1              False            False         False            False   
2              False            False         False            False   
3              False            False         False            False   
4              False            False         False            False   

   FuelType_Premium  FuelType_Unleaded  brand  Enginec  Cylinder  
0             False              False     56       14         3  
1              True              False     37        7         3  
2              True              False      4       12         3  
3              True              False     42       46         6  
4             False               True     51        5         3

# Dataset Ready

As visible from above, the Dataset which had all categorical values in different formats has been converted to numerical values and also sclaed to between 0 and 1 for the machine learning algorithms. 

# Splitting the Dataset

This Dataset will be split in the proportion of 70% for Training and 30% for testing. The Price column will be predicted and hence, the Price column is the Y .

In [113]:
X = cardf.drop(columns=['Price'])

y = cardf['Price']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Machine Learning

Since this Car Price prediction is a regression problem, The Following Machine Learning Algorithms will be tried and tested:

Linear regression,
Decision Tree,
Random Forest,
Gradient Boosting,

These algorithms will be Trained with a variety of different parameters.
The Model with the least Mean squared error and best R2 score will be selected as the one the best predictions.

In [114]:
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)
y_pred_lr = lr_model.predict(X_test)
rmse_lr = np.sqrt(mean_squared_error(y_test, y_pred_lr))
r2_lr = r2_score(y_test, y_pred_lr)
print(f"Linear Regression: RMSE = {rmse_lr}, R2 Score = {r2_lr}")

Linear Regression: RMSE = 11958.061785028793, R2 Score = 0.8462901626785428


In [115]:
dt_model = DecisionTreeRegressor(random_state=42)
dt_model.fit(X_train, y_train)
y_pred_dt = dt_model.predict(X_test)
rmse_dt = np.sqrt(mean_squared_error(y_test, y_pred_dt))
r2_dt = r2_score(y_test, y_pred_dt)
print(f"Decision Tree: RMSE = {rmse_dt}, R2 Score = {r2_dt}")

Decision Tree: RMSE = 12719.849642276124, R2 Score = 0.826082200337482


In [116]:
rf_model = RandomForestRegressor(random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)
rmse_rf = np.sqrt(mean_squared_error(y_test, y_pred_rf))
r2_rf = r2_score(y_test, y_pred_rf)
print(f"Random Forest: RMSE = {rmse_rf}, R2 Score = {r2_rf}")

Random Forest: RMSE = 8333.382583661583, R2 Score = 0.9253512193392236


In [117]:
gb_model = GradientBoostingRegressor(random_state=42)
gb_model.fit(X_train, y_train)
y_pred_gb = gb_model.predict(X_test)
rmse_gb = np.sqrt(mean_squared_error(y_test, y_pred_gb))
r2_gb = r2_score(y_test, y_pred_gb)
print(f"Gradient Boosting: RMSE = {rmse_gb}, R2 Score = {r2_gb}")

Gradient Boosting: RMSE = 9431.230407758905, R2 Score = 0.904387034844833


# Results

As Seen Above, the Model returns the following results:

R2 Score For Linear regression 0.84
R2 Score For Decision tree regression 0.82
R2 Score For Random Forest regression 0.925
R2 Score For Gradient Boosting regression 0.90

Since the Mean squared Error is the least for the Random Forest regression it can be said that this is the best performing Machine Learning Model for this dataset as it can predict used car prices with upto 92% accuracy and has the minimum Root Mean Squared Error of 8333. 

# Conclusion and Recommendations

This Dataset contains Valuable Information about the Different Used car Types and the prices in Australia.

The Machine Learning Pipeline is complete and as per results, The Random Forest regressor is the best model.

These reults can be futher made better by indulging in Hyper parameter tuning of different algorithms. 

However, there are certain Design Decisions that were taken to optimize the results and they need to be understood:

The Age and Kilometres Column were scaled in the Reverse order as The More the age and Kilometres, the less likely the Value of the Car is to be. 

The Interior and Exterior Columns Could be One Hot Encoded but that would significantly increase the dimensionality of the Data.

The No of Doors, Seats, Cylinders could be Label Encoded but it was chosen to keep their values as they help to understand the number attribute of the features. 